src/misc/unicode.cpp

```diff
@@ -0,0 +1,1129 @@
+{
+    error_parsing(file_name, line_num, "not currently defining a code page");
+}
+
+static bool check_import_status(const std::ifstream &in_file,
+                                const std::string &file_name,
+                                const bool empty)
+{
+    if (in_file.fail() && !in_file.eof()) {
+        LOG_ERR("UTF8: Error reading mapping file %s", file_name.c_str());
+        return false;
+    }
+
+    if (empty) {
+        LOG_ERR("UTF8: Mapping file %s has no entries", file_name.c_str());
+        return false;
+    }
+
+    return true;
+}
+
+static bool check_grapheme_valid(const Grapheme &grapheme,
+                                 const std::string &file_name,
+                                  const size_t line_num)
+{
+    if (grapheme.IsValid())
+        return true;
+
+    LOG_ERR("UTF8: Error, invalid grapheme defined in file %s, line %d",
+            file_name.c_str(), static_cast<int>(line_num));
+    return false;    
+}
+
+static const std::regex regex_map_ascii(
+    // hex_4_digits NNN|CPS|HSH|character
+    "^0x([[:xdigit:]]{4})[[:space:]]+.*(NNN|SPC|HSH|[^[:space:]])$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_code(
+    // hex_2_digits hex_4_digits
+    "^0x([[:xdigit:]]{2})[[:space:]]+0x([[:xdigit:]]{4})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_code_mark1(
+    // hex_2_digits hex_4_digits
+    "^0x([[:xdigit:]]{2})[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_code_mark2(
+    // hex_2_digits hex_4_digits
+    "^0x([[:xdigit:]]{2})[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_undefined(
+    // hex_2_digits hex_4_digits
+    "^0x[[:xdigit:]]{2}$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_alias(
+    // ALIAS hex_4_digits hex_4_digits [BIDIRECTIONAL]
+    "^ALIAS[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})([[:space:]]+BIDIRECTIONAL)?$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_cp_start(
+    // CODEPAGE dec
+    "^CODEPAGE[[:space:]]+([[:digit:]]{1,5})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_cp_duplicates(
+    // CODEPAGE dec DUPLICATES dec
+    "^CODEPAGE[[:space:]]+([[:digit:]]{1,5})[[:space:]]+DUPLICATES[[:space:]]+([[:digit:]]{1,5})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_extend_file(
+    // EXTENDS FILE dir_name file_NAME
+    "^EXTENDS[[:space:]]+FILE[[:space:]]+([^[:space:]]+)[[:space:]]+([^[:space:]]+)$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_extend_cp(
+    // EXTENDS CODEPAGE dec
+    "^EXTENDS[[:space:]]+CODEPAGE[[:space:]]+([[:digit:]]+)$",
+    std::regex_constants::optimize);
```

Comment:
I'm really not a fan of regexes, but if you really want to use them, I think they should be unit tested at least. It's very easy to make mistakes and not accept some valid input, or accept all sorts of invalid input, and/or miss or inappropriately handle edge cases, etc.

I've had so much trouble with regexes in the past that now if I use them, I always add a comprehensive unit test suite to them, so that's my recommendation. Most of the time you won't notice these bugs by just reading the regexes...

Reply:
Got rid of all the regexes.